Common/Util/BusyCollection.cs

```diff
@@ -0,0 +1,111 @@
+        private readonly ManualResetEventSlim _processingCompletedEvent = new ManualResetEventSlim(true);
+
+        /// <summary>
+        /// Gets a wait handle that can be used to wait until this instance is done
+        /// processing all of it's item
+        /// </summary>
+        public WaitHandle WaitHandle => _processingCompletedEvent.WaitHandle;
+
+        /// <summary>
+        /// Gets the number of items held within this collection
+        /// </summary>
+        public int Count => _collection.Count;
+
+        /// <summary>
+        /// Returns true if processing, false otherwise
+        /// </summary>
+        public bool IsBusy => _collection.Count > 0 || !_processingCompletedEvent.IsSet;
+
+        /// <summary>
+        /// Adds the items to this collection
+        /// </summary>
+        /// <param name="item">The item to be added</param>
+        public void Add(T item)
+        {
+            Add(item, CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Adds the items to this collection
+        /// </summary>
+        /// <param name="item">The item to be added</param>
+        /// <param name="cancellationToken">A cancellation token to observer</param>
+        public void Add(T item, CancellationToken cancellationToken)
+        {
+            // we're adding work to be done, mark us as busy
+            _processingCompletedEvent.Reset();
+            _collection.Enqueue(item);
+        }
+
+        /// <summary>
+        /// Provides a consuming enumerable for items in this collection.
+        /// </summary>
+        /// <returns>An enumerable that removes and returns items from the collection</returns>
+        public IEnumerable<T> GetConsumingEnumerable()
+        {
+            return GetConsumingEnumerable(CancellationToken.None);
+        }
+
+        /// <summary>
+        /// Provides a consuming enumerable for items in this collection.
+        /// </summary>
+        /// <param name="cancellationToken">A cancellation token to observer</param>
+        /// <returns>An enumerable that removes and returns items from the collection</returns>
+        public IEnumerable<T> GetConsumingEnumerable(CancellationToken cancellationToken)
+        {
+            T item;
+            while (_collection.TryDequeue(out item))
+            {
+                yield return item;
+            }
+            _processingCompletedEvent.Set();
+        }
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        /// <filterpriority>2</filterpriority>
+        public void Dispose()
+        {
+            _collection.Clear();
+            _processingCompletedEvent.Dispose();
+        }
+
+        /// <summary>
+        /// Marks the collection as not accepting any more additions
+        /// </summary>
+        public void CompleteAdding()
```

Comment:
In order to maintain behavior with the other implementation, subsequent calls to `Add` should fail after calling `CompleteAdding`.

See my good ol' friend [Liskov](https://en.wikipedia.org/wiki/Liskov_substitution_principle) for more information.